src/widget/textbox.rs

```diff
@@ -35,30 +44,153 @@ const BORDER_WIDTH: f64 = 1.;
+}
+
 #[derive(Debug, Clone)]
 pub struct TextBox {
     width: f64,
+    hscroll_offset: f64,
+    selection: Selection,
+    cursor_timer: TimerToken,
+    cursor_on: bool,
 }
 
 impl TextBox {
     pub fn new(width: f64) -> TextBox {
-        TextBox { width }
+        TextBox {
+            width,
+            hscroll_offset: 0.,
+            selection: Selection::caret(3),
+            cursor_timer: TimerToken::INVALID,
+            cursor_on: false,
+        }
     }
 
-    fn get_layout(
-        &mut self,
-        text: &mut <Piet as RenderContext>::Text,
-        font_size: f64,
-        data: &String,
-    ) -> <Piet as RenderContext>::TextLayout {
+    fn get_layout(&self, text: &mut PietText, font_size: f64, data: &String) -> PietTextLayout {
         // TODO: caching of both the format and the layout
         let font = text
-            .new_font_by_name("Roboto", font_size)
+            .new_font_by_name("Segoe UI", font_size)
             .unwrap()
             .build()
             .unwrap();
         text.new_text_layout(&font, data).unwrap().build().unwrap()
     }
+
+    fn insert(&mut self, src: &mut String, new: &str) {
+        // TODO: handle incomplete graphemes
+        let (left, right) = (self.selection.min(), self.selection.max());
+        src.replace_range(left..right, new);
+        self.selection = Selection::caret(left + new.len());
+    }
+
+    fn cursor_to(&mut self, to: usize) {
+        self.selection = Selection::caret(to);
+    }
+
+    fn cursor(&self) -> usize {
+        self.selection.end
+    }
+
+    /// Calculate a stateful scroll offset
+    fn update_hscroll(&mut self, rc_text: &mut PietText, data: &String) {
+        let cursor_x = self.substring_measurement_hack(rc_text, data, 0, self.cursor());
+
+        let padding = PADDING_LEFT * 2.;
+        if cursor_x > self.width + self.hscroll_offset - padding {
+            // If cursor goes past right side, bump the offset
+            //       ->
+            // **[****I]****
+            //   ^
+            self.hscroll_offset = cursor_x - self.width + padding;
+        } else if cursor_x < self.hscroll_offset {
+            // If cursor goes past left side, match the offset
+            //    <-
+            // **[I****]****
+            //   ^
+            self.hscroll_offset = cursor_x
+        }
+    }
+
+    // TODO: Grapheme isn't the correct unit for backspace, see:
+    // https://github.com/xi-editor/xi-editor/blob/master/rust/core-lib/src/backspace.rs
+    fn backspace(&mut self, src: &mut String) {
+        if self.selection.is_caret() {
+            let cursor = self.cursor();
+            let new_cursor = prev_grapheme(&src, cursor);
+            src.replace_range(new_cursor..cursor, "");
+            self.cursor_to(new_cursor);
+        } else {
+            src.replace_range(self.selection.min()..self.selection.max(), "");
+            self.cursor_to(self.selection.min());
+        }
+    }
+
+    fn copy_text(&self, input: String) {
```

Comment:
maybe throw in a `TODO` note here, explaining you're waiting for copy/past to get implemented?